### PR TITLE
Physics units: Fixes wrong dimension calculations for functions.

### LINKED ISSUE #20288: Dimensionless number in exponent not recognized as dimensionless in _collect_factor_and_dimension
The following raises an error:
```python
from sympy import S, exp
from sympy.physics.units import energy, length, mass, time, Quantity
u = Quantity("u")
v = Quantity("v")
SI.set_quantity_dimension(u, energy)
SI.set_quantity_dimension(v, energy)
Quantity._collect_factor_and_dimension(1 + exp(u/v))
```
returns:
`ValueError: Dimension of "exp(u/w)" is exp(Dimension(1)), but it should be Dimension(1)`

Using `check_dimensions` does not raise an error here:
```python
from sympy.physics.units.util import convert_to, check_dimensions
check_dimensions(1 + exp(u/v))
```

### PR DESCRIPTION
**Pull request by faze-geek**
#### References to other Issues or PRs

Fixes #20288
Closes #20333

#### Brief description of what is fixed or changed
Files Modified :
`sympy/physics/units/tests/test_quantities.py`
`sympy/physics/units/unitsystem.py
`

The correction now handles the dimension calculation of expressions with particular functions like` log` , `exp` and multiple functions of such kind too .

Currently on master:
```
>>> SI.set_quantity_dimension(u, energy)
>>> SI.set_quantity_dimension(v, energy)
>>> SI._collect_factor_and_dimension(1 + exp(u/v))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\kunni\sympy\sympy\sympy\physics\units\unitsystem.py", line 179, in _collect_factor_and_dimension
    raise ValueError(
ValueError: Dimension of "exp(u/v)" is exp(Dimension(1)), but it should be Dimension(1)
```
In Branch : 
```
>>> SI.set_quantity_dimension(u, energy)
>>> SI.set_quantity_dimension(v, energy)
>>> SI._collect_factor_and_dimension(1 + exp(u/v))
(1 + E, Dimension(1))
```
#### Other comments
Ansh Mishra ( @theanshm ) has worked on this issue quite a while ago, so I wouldn't expect him to revert for reviews now . He has been made author of the commit  https://github.com/sympy/sympy/pull/23296/commits/203771de6c61280da9934edfc40682a4383e5891 .

#### Release Notes

* physics.units
  * Fixed wrong dimension calculations for some mathematical functions.

### CODE REVIEW

File: sympy/physics/units/tests/test_quantities.py [hunk 1]
```diff
@@ -528,6 +528,20 @@ def test_issue_22819():
...
+def test_issue_20288():
+    from sympy import sympify
+    from sympy.core.numbers import E
+    from sympy.physics.units import energy
+    u = Quantity('u')
+    v = Quantity('v')
+    SI.set_quantity_dimension(u, energy)
+    SI.set_quantity_dimension(v, energy)
+    u.set_global_relative_scale_factor(1, energy)
+    v.set_global_relative_scale_factor(1, energy)
```
Comment (Upabjojr) [inline sympy/physics/units/tests/test_quantities.py:540]: The scale factor is supposed to be set relative to another quantity object, not to a dimension. `energy` is a `Dimension`.

Comment (faze-geek) [reply sympy/physics/units/tests/test_quantities.py:540]: > `energy` is a `Dimension`.

Then would something like this suffice .
```
>>> u.set_global_relative_scale_factor(1, joule)
>>> v.set_global_relative_scale_factor(1, joule)
```
Joule , Calories etc. would be considered as quantities of energy right ?

Comment (faze-geek) [reply sympy/physics/units/tests/test_quantities.py:540]: I guess this is correct , I make a commit as soon as I get a green flag !

Comment (faze-geek) [reply sympy/physics/units/tests/test_quantities.py]: @Upabjojr I have made the change , would appreciate a review !

Comment (Upabjojr) [reply sympy/physics/units/tests/test_quantities.py]: Yes, `joule` is a quantity corresponding to the dimension of `energy` (in every unit/dimension system).

Comment (faze-geek) [reply sympy/physics/units/tests/test_quantities.py]: Thanks, I have made the change so you can check it out .Does the pr look good to go then ?

File: sympy/physics/units/tests/test_quantities.py [hunk 2]
```diff
@@ -528,6 +528,20 @@ def test_issue_22819():
...
+    from sympy.core.numbers import E
+    from sympy.physics.units import energy
+    u = Quantity('u')
+    v = Quantity('v')
+    SI.set_quantity_dimension(u, energy)
+    SI.set_quantity_dimension(v, energy)
+    u.set_global_relative_scale_factor(1, joule)
+    v.set_global_relative_scale_factor(1, joule)
+    expr = 1 + exp(u**2/v**2)
+    assert SI._collect_factor_and_dimension(expr) == (sympify(1 + E), Dimension(1))
```
Comment (Upabjojr) [inline sympy/physics/units/tests/test_quantities.py:542]: why do you need to sympify this expression?

Comment (faze-geek) [reply sympy/physics/units/tests/test_quantities.py]: It was actually unnecessary, I tried to mimic a test . Thanks for pointing it out ! It has been removed in the latest commit .

**Comment by Upabjojr** [comment]
I see you are resuming another PR. Can you amend commit [203771d](https://github.com/sympy/sympy/pull/23296/commits/203771de6c61280da9934edfc40682a4383e5891) to have the authorship of the original author?

**Comment by faze-geek** [comment]
> I see you are resuming another PR. 

Yes , had mentioned in other comments (OP).
>Can you amend commit [203771d](https://github.com/sympy/sympy/pull/23296/commits/203771de6c61280da9934edfc40682a4383e5891) to have the authorship of the original author?

I haven't revived pr's before so I'm not aware of how this is done . Could you assist me here ? I will try to figure it out from online literature till then . Thanks !

**Comment by Upabjojr** [comment]
> I haven't revived pr's before so I'm not aware of how this is done . Could you assist me here ?

https://confluence.atlassian.com/bitbucketserverkb/how-do-you-make-changes-on-a-specific-commit-779171729.html

That is, use git rebase up to the parent commit, then in the git rebase editor change the commit you want to modify to _edit_ (all the other ones remain in _pick_ mode). At this point use `git commit --amend --author="..."`

**Comment by Upabjojr** [comment]
After rebasing, you probably need to force-push the changes to GitHub (ordinary push will probably fail because you are modifying the history of commits).

**Comment by faze-geek** [comment]
> That is, use git rebase up to the parent commit, then in the git rebase editor change the commit you want to modify to _edit_ (all the other ones remain in _pick_ mode). At this point use `git commit --amend --author="..."`

Yes I have used rebase multiple times before , I'll try the method you have suggested soon .

**Comment by faze-geek** [comment]
@Upabjojr 
I guess I have followed the steps correctly . Could you check it out !
Command -
`git commit --amend --author="Ansh Mishra <anshmishra471@gmail.com>"`

Information of commit -
```
[detached HEAD 66209bf28a] Fix_for_20288
 Author: Ansh Mishra <anshmishra471@gmail.com>
 Date: Sun Mar 27 16:51:58 2022 +0530
 2 files changed, 15 insertions(+), 1 deletion(-)
```

The only doubt I had was that force pushing did not re run tests as I have observed before -
```
(sympy-dev) C:\Users\kunni\sympy\sympy>git push --force origin dimension_fix
Enter passphrase for key '/c/Users/kunni/.ssh/id_rsa':
Everything up-to-date
```

**Comment by faze-geek** [comment]
All the tests have passed now . I believe I have answered the set of reviews . Is it good to go in?

**Comment by Upabjojr** [comment]
@faze-geek thanks for your contribution!

**Pull request merged**